src/main.rs

```diff
@@ -175,24 +212,101 @@ fn generate_solid_background(color: Color, background_surface: &mut WlSurface,
                              env: &WaylandEnv) -> BufferResult {
     // TODO support more formats, split into separate function
-    let image = open(path)
+    let mut image = open(path)
         .unwrap_or_else(|_| {
             println!("Could not open image file \"{:?}\"", path);
             ::std::process::exit(1);
         });
     /*let image = load_from_memory(CURSOR)
         .expect("Could not read cursor data, report to maintainer!");*/
-    let mut image = image.to_rgba();
-    let width = image.width();
-    let height = image.height();
-    let stride = width * 4;
-    let size = stride * height;
+    let mut img_width = image.width();
+    let mut img_height = image.height();
+
+    let dbus_con = Connection::get_private(BusType::Session).unwrap();
+    let resolution = get_screen_resolution(dbus_con);
+    let (scr_width, scr_height) = (resolution.0 as u32, resolution.1 as u32);
+
+    // Mode image processing
+    // The output must be scr_width x scr_height resolution
+    image = match mode {
+        BackgroundMode::Fill    => {
+            // Find fit scale
+            let width_sr: f64  = scr_width as f64 / img_width as f64;
+            let height_sr: f64 = scr_height as f64 / img_height as f64;
+            let scale_ratio: f64 = if width_sr > height_sr {
+                width_sr
+            } else {
+                height_sr
+            };
+            img_width = (scale_ratio * img_width as f64) as u32;
+            img_height = (scale_ratio * img_height as f64) as u32;
+
+            image = image.resize(img_width, img_height, FilterType::Gaussian);
+            image.crop(((img_width as i32 - scr_width as i32) / 2).abs() as u32,
+                ((img_height as i32 - scr_height as i32) / 2).abs() as u32,
+                scr_width,
+                scr_height)
+        },
+        // TODO: Padding background color
+        BackgroundMode::Fit     => {
+            // Find fit scale ratio
+            let width_sr: f64  = scr_width as f64 / img_width as f64;
+            let height_sr: f64 = scr_height as f64 / img_height as f64;
+            let scale_ratio: f64 = if width_sr < height_sr {
+                width_sr
+            } else {
+                height_sr
+            };
+            img_width = (scale_ratio * img_width as f64) as u32;
+            img_height = (scale_ratio * img_height as f64) as u32;
+
+            image = image.resize(img_width, img_height, FilterType::Gaussian);
+
+            let mut imagepad = DynamicImage::new_rgba8(scr_width, scr_height);
+            imagepad.copy_from(&image, 0, ((scr_height - img_height) / 2) as u32);
+
+            imagepad
+        },
+        BackgroundMode::Stretch => {
+            image.resize_exact(scr_width, scr_height, FilterType::Gaussian)
+        },
+        // TODO: Padding background color
+        BackgroundMode::Center  => {
+            // FIXME: How if image bigger than screen size?
```

Comment:
To handle that case, you'd want to use `sub_image`.

It seems possible to combine all 4 cases `(<W,<H), (>W,<H), (<W,>H), (>W, >H)` by using `sub_image` and `copy_from` with the appropriate mix of `max(difference,0)` and `max(-difference,0)`.